aws/sdk/examples/telephone-game/src/main.rs

```diff
@@ -0,0 +1,432 @@
+                    sink.append(source);
+                    sink.sleep_until_end();
+
+                    info!("Did you hear it?");
+                }
+                Err(e) => error!("{}", e),
+            };
+
+            // Clear last iteration's line so we can receive new user input
+            line.clear();
+        }
+    })
+}
+
+/**
+# Play a game of Telephone w/ AWS
+
+1. Fetch user any user input that will override default values
+1. Create a config and required clients for AWS services
+1. Create a bucket to store audio and transcriptions if none exists
+1. Run in a loop
+    1. Start a speech synthesis task and set it to output to the previously created S3 bucket
+    1. Wait up to TASK_TIMEOUT_IN_SECONDS seconds for synthesis task to complete
+        - The status of the task is checked every TASK_WAIT_INTERVAL_IN_SECONDS in a loop
+        - Break out of the loop once the task succeeds or fails
+    1. Delete any pre-existing transcription job with the name "telephone-game-transcription"
+        - Job names must be unique so we clear the old job to reuse the name.
+    1. Wait up to TASK_TIMEOUT_IN_SECONDS seconds for transcripion job to complete
+        - The status of the job is checked every TASK_WAIT_INTERVAL_IN_SECONDS in a loop
+        - Break out of the loop once the job succeeds or fails
+    1. Download the transcript JSON file from S3 and print what Transcribe heard
+    1. Break out of the loop if an error occurs or we reach the iteration limit
+1. Log a final message before closing
+    - If everything succeeded, log the original phrase, the current phrase, and how many iterations occurred
+    - If a failure occurs, log the error and any errors that caused it
+*/
+fn play_telephone(matches: &ArgMatches) {
+    let rt = Runtime::new().expect("failed to create an async runtime");
+    let number_of_iterations = matches
+        .value_of("iterations")
+        .and_then(|i| i.parse::<u32>().ok())
+        .unwrap_or(DEFAULT_NUMBER_OF_ITERATIONS);
+
+    if number_of_iterations == 0 {
+        error!("Iterations must be a number greater than 0");
+        return;
+    }
+
+    let original_phrase = matches.value_of("phrase").unwrap_or_default();
+    let mut current_phrase = original_phrase.to_owned();
+
+    let bucket_name = matches
+        .value_of("s3_bucket_name")
+        .map(ToOwned::to_owned)
+        .unwrap_or_else(|| "telephone-game".to_owned());
+
+    let result = rt.block_on(async {
+        let config = aws_config::load_from_env().await;
+        let s3_client = aws_sdk_s3::Client::new(&config);
+        let polly_client = aws_sdk_polly::Client::new(&config);
+        let transcribe_client = aws_sdk_transcribe::Client::new(&config);
+        let bucket_name = create_s3_bucket_if_not_exists(&s3_client, &bucket_name)
+            .await
+            .context("Failed to complete necessary setup")?;
+
+        let mut i = 0;
+        'game: loop {
```

Comment:
Consider breaking this into a few different async functions.

Reply:
Is the rationale for this that it's easier to read stuff if it's not indented?

Reply:
In general, it's easier to reason about code when functions are kept below about 50-60 lines. But in this case, having separate functions for the different steps would make the loop read like a step by step overview of the process, which may reduce the necessity of the comment on top of the function as comments could be sprinkled along each step. It's also useful to be able to see all the interactions with the loop index on one screen, and to be able to quickly reason about all the exit points from the loop.